### PR TITLE
update SDP cone definition to use lower triangle

### PR DESCRIPTION
**Pull request by mlubin**
CC @karanveerm @madeleineudell @ulfworsoe @bodono

### CODE REVIEW

**Comment by ulfworsoe** [comment]
I will update Mosek.jl as soon as the patch is applied.

**Comment by mlubin** [comment]
Thanks @ulfworsoe.
@madeleineudell @karanveerm, I'm waiting for the okay from the Convex.jl team.

File: doc/conic.rst [hunk 1]
```diff
@@ -53,14 +53,16 @@ Not all solvers are expected to support all types of cones. However, when a simp
...
-    must be a square integer :math:`n` corresponding to a
-    :math:`\sqrt{n}\times \sqrt{n}` matrix;
-    variables should be listed in column-major, or by symmetry,
-    row-major order.
+    For the semidefinite cone, the number of variables or constraints
+    present correspond to the lower (or upper) triangular elements
+    in column-major (resp., row-major) order. Since an :math:`n \times n`
+    matrix has :math:`\frac{n(n+1)}{2}` lower-triangular elements,
+    by inverting this formula, when :math:`y` elements are specified in
+    ``indices``, the corresponding matrix is
```
Comment (karanveerm) [inline doc/conic.rst]: Shouldn't this be worded as, the corresponding matrix has :math:\`\left(\sqrt{\frac{1}{4}+2y}-\frac{1}{2}\right) \times \left(\sqrt{\frac{1}{4}+2y}-\frac{1}{2}\right)\` elements?

**Comment by karanveerm** [comment]
I think this is fine. 

Should MPB conic interface also perhaps support sdp cones with n^2 elements, and take care of converting these to n(n+1)/2 elements?

**Comment by mlubin** [comment]
Thanks for the comments, @karanveerm.

> Should MPB conic interface also perhaps support sdp cones with n^2 elements, and take care of converting these to n(n+1)/2 elements?

I'd prefer to avoid this at least until everything is up and running with the new triangular interface. If we allow n^2 elements we introduce extra work for the solver interfaces, and we have to provide a very clear mathematical definition of what it means if the input is not symmetric.

**Comment by madeleineudell** [comment]
I'd vote for (A + A') / 2 in case of asymmetry, but it's also fine to just
not allow it for now. Otherwise we'd need to identify when we're in either
case, which leaves more room for error.

On Mon, Feb 2, 2015 at 2:42 PM, Miles Lubin <notifications@github.com>
wrote:

> Thanks for the comments, @karanveerm <https://github.com/karanveerm>.
>
> Should MPB conic interface also perhaps support sdp cones with n^2
> elements, and take care of converting these to n(n+1)/2 elements?
>
> I'd prefer to avoid this at least until everything is up and running with
> the new triangular interface. If we allow n^2 elements we introduce extra
> work for the solver interfaces, and we have to provide a very clear
> mathematical definition of what it means if the input is not symmetric.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/JuliaOpt/MathProgBase.jl/pull/55#issuecomment-72554482>
> .
>



-- 
Madeleine Udell
PhD Candidate in Computational and Mathematical Engineering
Stanford University
www.stanford.edu/~udell

**Comment by mlubin** [comment]
@madeleineudell, do you mean that you're happy with the triangular form as is or that we should add an interface that accepts square SDP terms which must be symmetric?

**Comment by madeleineudell** [comment]
yes, I'm happy with the triangular form. sorry for dragging on this; we've
been engaged in long discussions off github over the true meaning of dual
values for SDPs...

On Mon, Feb 2, 2015 at 5:20 PM, Miles Lubin <notifications@github.com>
wrote:

> @madeleineudell <https://github.com/madeleineudell>, do you mean that
> you're happy with the triangular form as is or that we should add an
> interface that accepts square SDP terms which must be symmetric?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/JuliaOpt/MathProgBase.jl/pull/55#issuecomment-72574331>
> .
>



-- 
Madeleine Udell
PhD Candidate in Computational and Mathematical Engineering
Stanford University
www.stanford.edu/~udell

**Comment by bodono** [comment]
So the question is what to do when the user wants to solve something with constraints:

Bx + vec(S) == b
S == semidefinite(n) // of dimension n*n

where S is a variable of size n^2. One possibility is something like: 

Bx + vec(S) == b
lower(S) == lower(U)
lower(S^T) == lower(U)
U == semidefinite(n) // of dimension n*(n+1)/2

that way we don't touch the B matrix, and the problem is a valid SDP in the new format. There are other equivalent ways to formulate it too, which may be easier to implement in MPB / Conjex.jl etc.

**Comment by ulfworsoe** [comment]
Just nitpicking: You probably meant 
```
:math:`\left(\sqrt{\frac{1}{4}+2y}-\frac{1}{2}\right) \times \left(\sqrt{\frac{1}{4}+2y}+\frac{1}{2}\right)`
```
in line 64 in  conic.rst ('+' instead of '-' in second parenthesis).

@mlubin

**Pull request merged**